### PR TITLE
fix: fixed test suite and baseapp issues from cherry-pick

### PR DESCRIPTION
**Pull request by Vvaradinov**
## Description
This PR fixes some merge conflicts, updates the unit tests and cleans up `baseapp` with the 3 `PostHandler` cases

Closes: #XXXX

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: baseapp/baseapp.go [hunk 1]
```diff
@@ -6,17 +6,16 @@ import (
 	"sort"
 	"strings"
 
-	dbm "github.com/cosmos/cosmos-db"
 	"github.com/cosmos/gogoproto/proto"
 	abci "github.com/tendermint/tendermint/abci/types"
 	"github.com/tendermint/tendermint/crypto/tmhash"
 	"github.com/tendermint/tendermint/libs/log"
 	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
+	dbm "github.com/tendermint/tm-db"
```
Comment (fedekunze) [inline baseapp/baseapp.go]: curious why is this change needed?

File: baseapp/baseapp.go [hunk 2]
```diff
@@ -718,35 +722,83 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte) (gInfo sdk.GasInfo, re
...
-		// Run optional postHandlers.
-		//
-		// Note: If the postHandler fails, we also revert the runMsgs state.
-		if app.postHandler != nil {
-			// Follow-up Ref: https://github.com/cosmos/cosmos-sdk/pull/13941
-			newCtx, err := app.postHandler(runMsgCtx, tx, mode == runTxModeSimulate, err == nil)
-			if err != nil {
-				return gInfo, nil, anteEvents, priority, err
-			}
+	// Case 2: Run PostHandler and only
```
Comment (fedekunze) [inline baseapp/baseapp.go]: finish comment

Comment (fedekunze) [reply baseapp/baseapp.go]:
```suggestion
	// Case 2: tx errors and the post handler is set. Run PostHandler and revert state from runMsgs
```

File: baseapp/baseapp.go [hunk 3]
```diff
@@ -718,35 +722,83 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte) (gInfo sdk.GasInfo, re
...
+		} else {
+			events = make([]abci.Event, len(postEvents))
+			copy(events, postEvents)
+			result.Events = append(result.Events, postEvents...)
 		}
+
+		return gInfo, result, events, priority, err
+	}
+
+	// Case 3: Run Post Handler with runMsgCtx so that the state from runMsgs is persisted
```
Comment (fedekunze) [inline baseapp/baseapp.go]:
```suggestion
	// Case 3: tx successful and post handler is set. Run Post Handler with runMsgCtx so that the state from runMsgs is persisted
```

File: baseapp/baseapp.go [hunk 4]
```diff
@@ -718,35 +722,83 @@ func (app *BaseApp) runTx(mode runTxMode, txBytes []byte) (gInfo sdk.GasInfo, re
...
+		if len(anteEvents) > 0 {
+			events = make([]abci.Event, len(anteEvents)+len(postEvents))
+			copy(events[:len(anteEvents)], anteEvents)
+			copy(events[len(anteEvents):], postEvents)
+		} else {
+			events = make([]abci.Event, len(postEvents))
+			copy(events, postEvents)
+		}
+	}
+
```
Comment (fedekunze) [inline baseapp/baseapp.go:784]: I'd write a comment about case 4 (tx successful and post handler not set)

File: baseapp/baseapp_test.go [hunk 5]
```diff
@@ -1,8 +1,11 @@
 package baseapp_test
 
 import (
+	"cosmossdk.io/depinject"
```
Comment (fedekunze) [inline baseapp/baseapp_test.go]: fix linter

File: baseapp/baseapp.go (hunk 1 above)
Comment (Vvaradinov) [reply baseapp/baseapp.go]: Fixed this, had to merge the latest from main

File: CHANGELOG.md [hunk 6]
```diff
@@ -126,6 +126,7 @@ Ref: https://keepachangelog.com/en/1.0.0/
 
 ### State Machine Breaking
 
+* (baseapp, x/auth/posthandler) [#13940](https://github.com/cosmos/cosmos-sdk/pull/13940) Update `PostHandler` to receive the `runTx` call `Result` and success boolean.
```
Comment (fedekunze) [inline CHANGELOG.md]:
```suggestion
* (baseapp, x/auth/posthandler) [#13940](https://github.com/cosmos/cosmos-sdk/pull/13940) Update `PostHandler` to receive the `runTx` `success` boolean.
```

**Review by fedekunze — APPROVED** [review]
ACK. We need to update the changelog entry because we're making the post handler stateful as well
